### PR TITLE
Show component versions on hw detail

### PR DESCRIPTION
**Pull request by caberos**
https://github.com/softlayer/softlayer-python/issues/1465

### CODE REVIEW

File: SoftLayer/CLI/hardware/detail.py [hunk 1]
```diff
@@ -107,6 +108,46 @@ def cli(env, identifier, passwords, price):
             pass_table.add_row([item['username'], item['password']])
         table.add_row(['remote users', pass_table])
 
+    if components:
+
+        _mask = 'id,hardwareComponentModel[name,' \
```
Comment (ATGE) [inline SoftLayer/CLI/hardware/detail.py:113]: moving `_mask` and `_filter` logic within the `getcomponents` , will look better.

File: SoftLayer/CLI/hardware/detail.py [hunk 2]
```diff
@@ -107,6 +108,46 @@ def cli(env, identifier, passwords, price):
...
+                                    "DESC"
+                                ]
+                            },
+                            {
+                                "name": "sortOrder",
+                                "value": [
+                                    1
+                                ]}]}
+                }}}}
+        components = hardware.get_components(identifier, _mask, _filter)
```
Comment (ATGE) [inline SoftLayer/CLI/hardware/detail.py:135]:
```suggestion
        components = hardware.get_components(identifier)
```
after moving `_mask` and `_filter` within `get_components(identifier)`

File: tests/managers/hardware_tests.py [hunk 3]
```diff
@@ -897,6 +897,22 @@ def test_upgrade_full(self):
...
+                       'serviceProviderId': 1,
+                       'hardwareComponentModel':
+                           {'name': 'IMM2 - Onboard',
+                            'firmwares':
+                                [{'createDate': '2020-09-24T13:46:29-06:00',
+                                  'version': '5.60'},
+                                 {'createDate': '2019-10-14T16:51:12-06:00',
+                                  'version': '5.10'}]}}]
+        self.assertEqual(result, components)
+
```
Comment (ATGE) [inline tests/managers/hardware_tests.py:915]:
```suggestion
self.assert_called_with('SoftLayer_Hardware_Server',
                                'getComponents',
                                identifier=1234)
```
testing the API calls helps to sure they will work as expected too.

File: SoftLayer/CLI/hardware/detail.py [hunk 4]
```diff
@@ -107,6 +110,24 @@ def cli(env, identifier, passwords, price):
             pass_table.add_row([item['username'], item['password']])
         table.add_row(['remote users', pass_table])
 
+    if components:
+        components = hardware.get_components(identifier)
+        components_table = formatting.Table(['name', 'version', 'date', 'component-type'])
```
Comment (ATGE) [inline SoftLayer/CLI/hardware/detail.py:115]:
```suggestion
                components_table = formatting.Table(['name', 'Firmware version', 'Firmware build date', 'Type'])
```
maybe adding the id too

File: SoftLayer/CLI/hardware/detail.py [hunk 5]
```diff
@@ -107,6 +110,24 @@ def cli(env, identifier, passwords, price):
...
+        components_table.align['date'] = 'l'
+        ids = []
+        for component in components:
+            if component['id'] not in ids:
+                firwmare = component['hardwareComponentModel']['firmwares'][0]
+                components_table.add_row([utils.lookup(component, 'hardwareComponentModel', 'name'),
```
Comment (ATGE) [inline SoftLayer/CLI/hardware/detail.py:121]:
```suggestion
               components_table.add_row([utils.lookup(component, 'hardwareComponentModel', 'longDescription'),
```

**Review by ATGE — COMMENTED** [review]
### Testing
- ✔️ slcli hw detail --components 1866407 
```bash
:       components : :.....................:................:...........................:..................:                         :
:                  : :         name        :    version     : date                      :  component-type  :                         :
:                  : :.....................:................:...........................:..................:                         :
:                  : :    X11DPU+_R1.10    : 3.3a 7-21-2020 : 2020-08-28T12:48:59-06:00 :   MOTHERBOARD    :                         :
:                  : :   AOC-2UR66-i4XTF   :   6.01(6.01)   : 2021-02-16T11:46:32-06:00 :   NETWORK_CARD   :                         :
:                  : :       9461-16i      : 5.130.00-3059  : 2020-05-22T10:44:07-06:00 : DRIVE_CONTROLLER :                         :
:                  : :  AST2500 - Onboard  :      3.10      : 2020-09-16T15:09:26-06:00 : REMOTE_MGMT_CARD :                         :
:                  : :    AOM-TPM-9671H    :   6.43.243.0   : 2020-01-24T15:28:41-06:00 : SECURITY_DEVICE  :                         :
:                  : : Enterprise Capacity :   TN04 TN04    : 2019-07-09T20:58:53-06:00 :    HARD_DRIVE    :                         :
:                  : : Enterprise Capacity :   TN04 TN04    : 2019-07-09T20:58:53-06:00 :    HARD_DRIVE    :                         :
:                  : :.....................:................:...........................:..................:                         :
:             tags : -                                                                                                               :
:..................:.................................................................................................................:
```
below, some suggestions to improve the information about the components(maybe adding the id too)::

```
:       components : :...............................................................................:..................:...........................:..................: :
:                  : :                                      name                                     : Firmware version :    Firmware build date    :       Type       : :
:                  : :...............................................................................:..................:...........................:..................: :
:                  : :       SuperMicro / MBD-X11DPU / X11DPU+_R1.10 / Intel Xeon DualProc / 2       :  3.3a 7-21-2020  : 2020-08-28T12:48:59-06:00 :   MOTHERBOARD    : :
:                  : :             SuperMicro / AOC-2UR66-I4XTF / AOC-2UR66-i4XTF / 1 / 4            :    6.01(6.01)    : 2021-02-16T11:46:32-06:00 :   NETWORK_CARD   : :
:                  : : LSI / DRIVE CONTROLLER / 9461-16i / SATA/SAS - MegaRAID SAS 9461-16i SGL / 16 :  5.130.00-3059   : 2020-05-22T10:44:07-06:00 : DRIVE_CONTROLLER : :
:                  : :       Aspeed / AST2500 - Onboard / IPMI - KVM / Remote Management Count1      :       3.10       : 2020-09-16T15:09:26-06:00 : REMOTE_MGMT_CARD : :
:                  : :     SuperMicro / AOM-TPM-9671H-P / AOM-TPM-9671H / Infineon TPM Module / 1    :    6.43.243.0    : 2020-01-24T15:28:41-06:00 : SECURITY_DEVICE  : :
:                  : :       Seagate / ST1000NM0055 / Enterprise Capacity / ST1000NM0055 / 1000      :    TN04 TN04     : 2019-07-09T20:58:53-06:00 :    HARD_DRIVE    : :
:                  : :       Seagate / ST1000NM0055 / Enterprise Capacity / ST1000NM0055 / 1000      :    TN04 TN04     : 2019-07-09T20:58:53-06:00 :    HARD_DRIVE    : :
:                  : :...............................................................................:..................:...........................:..................: :
:             tags : -                                                                                                                                                   :
:..................:.....................................................................................................................................................:

```

File: SoftLayer/CLI/hardware/detail.py [hunk 6]
```diff
@@ -9,13 +9,16 @@
...
 from SoftLayer import utils
 
+# pylint: disable=R0915
+
 
 @click.command()
 @click.argument('identifier')
 @click.option('--passwords', is_flag=True, help='Show passwords (check over your shoulder!)')
 @click.option('--price', is_flag=True, help='Show associated prices')
+@click.option('--components', is_flag=True, help='Show associated hardware components')
```
Comment (allmightyspiff) [inline SoftLayer/CLI/hardware/detail.py:19]: Set the default to false

File: SoftLayer/CLI/hardware/detail.py [hunk 7]
```diff
@@ -107,6 +110,24 @@ def cli(env, identifier, passwords, price):
             pass_table.add_row([item['username'], item['password']])
         table.add_row(['remote users', pass_table])
 
+    if components:
+        components = hardware.get_components(identifier)
+        components_table = formatting.Table(['name', 'version', 'date', 'component-type'])
+        components_table.align['date'] = 'l'
+        ids = []
+        for component in components:
+            if component['id'] not in ids:
```
Comment (allmightyspiff) [inline SoftLayer/CLI/hardware/detail.py:119]: Are there duplicate `ids` in the components list? I would assume this check wouldn't be needed. But if this check is preventing duplicate line entries or something feel free to keep it in there.

**Review by allmightyspiff — CHANGES_REQUESTED** [review]
Looks good. I like @atge 's suggestions, once those are added this should be good to go.

Thanks.

**Review by allmightyspiff — CHANGES_REQUESTED** [review]
Looks good, but there are some unintended changes here. Please update this branch with the latest changes from master and I think those should go away, hopefully.

Once that is done this should be good to go. Thanks.

**Comment by allmightyspiff** [comment]
Thanks for merging in the changes from master, but it still looks like your accidently committing changes you didn't mean to.
Take a look at https://github.com/softlayer/softlayer-python/pull/1470/files and make sure ONLY the changes you intent to make are in there.

**Review by allmightyspiff — APPROVED** [review]

**Pull request merged**